### PR TITLE
[BEAM-775] Remove Aggregators from the Java SDK

### PR DESCRIPTION
**Pull request by pabloem**
Removing from internal SDK usage and DoFn contexts so users can not create them.
Some failures in Apex code remain. All else seems to be `mvn clean verify`ing well.

### CODE REVIEW

**Comment by pabloem** [comment]
@bjchambers can you trigger a Jenkins run?

**Comment by pabloem** [comment]
Apex runner issue fixed. Changes pass `mvn clean verify`

**Comment by pabloem** [comment]
r: @dhalperi

File: runners/core-java/src/main/java/org/apache/beam/runners/core/DoFnAdapters.java [hunk 1]
```diff
@@ -72,7 +69,7 @@ private DoFnAdapters() {}
     private transient DoFnInvoker<InputT, OutputT> invoker;
 
     SimpleDoFnAdapter(DoFn<InputT, OutputT> fn) {
-      super(AggregatorRetriever.getDelegatingAggregators(fn));
+      super();
```
Comment (dhalperi) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/DoFnAdapters.java]: drop this line - it is now a no-op.

File: runners/apex/src/main/java/org/apache/beam/runners/apex/translation/operators/ApexGroupByKeyOperator.java [hunk 2]
```diff
@@ -197,7 +198,7 @@ public void outputWindowedValue(
           }
         },
         NullSideInputReader.empty(),
-        null,
```
Comment (dhalperi) [inline runners/apex/src/main/java/org/apache/beam/runners/apex/translation/operators/ApexGroupByKeyOperator.java]: Can you say more about this change? Is this a bug in Apex? Does not seem Aggregator-related.

Comment (dhalperi) [inline runners/apex/src/main/java/org/apache/beam/runners/apex/translation/operators/ApexGroupByKeyOperator.java]: Note to self -- it's simply a bug fixup on prior CL and does not make sense in isolation.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/GroupAlsoByWindowViaOutputBufferDoFn.java [hunk 3]
```diff
@@ -70,7 +70,6 @@ public void processElement(ProcessContext c) throws Exception {
             timerInternals,
             WindowingInternalsAdapters.outputWindowedValue(c.windowingInternals()),
             WindowingInternalsAdapters.sideInputReader(c.windowingInternals()),
-            droppedDueToClosedWindow,
```
Comment (dhalperi) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/GroupAlsoByWindowViaOutputBufferDoFn.java]: Note to self: okay, inherited from parent abstract class

File: runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java [hunk 4]
```diff
@@ -72,7 +73,8 @@
  *
  * <ul>
  * <li>Tracking the windows that are active (have buffered data) as elements arrive and triggers are
- *     fired.
+ *
```
Comment (dhalperi) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java]: revert the new line added here?

File: runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java [hunk 5]
```diff
@@ -109,7 +111,9 @@
 
   private final StateInternals<K> stateInternals;
 
-  private final Aggregator<Long, Long> droppedDueToClosedWindow;
+  private final Counter droppedDueToClosedWindow;
+
+  public static final String DROPED_DUE_TO_CLOSED_WINDOW = "droppedDueToClosedWindow";
```
Comment (dhalperi) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java]: typo: DROPPED

File: sdks/java/core/src/main/java/org/apache/beam/sdk/AggregatorPipelineExtractor.java [hunk 6]
```diff
@@ -25,7 +25,6 @@
 import org.apache.beam.sdk.Pipeline.PipelineVisitor;
 import org.apache.beam.sdk.runners.TransformHierarchy;
 import org.apache.beam.sdk.transforms.Aggregator;
-import org.apache.beam.sdk.transforms.AggregatorRetriever;
 import org.apache.beam.sdk.transforms.PTransform;
 import org.apache.beam.sdk.transforms.ParDo;
 import org.apache.beam.sdk.values.PValue;
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/AggregatorPipelineExtractor.java]: Is this file still needed? If so, please `@Deprecated` the main class and comment why.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java [hunk 7]
```diff
@@ -699,66 +637,11 @@ public Duration getAllowedTimestampSkew() {
...
-  }
-
-  /**
    * Finalize the {@link DoFn} construction to prepare for processing.
    * This method should be called by runners before any processing methods.
    */
   @Deprecated
   public final void prepareForProcessing() {
-    aggregatorsAreFinal = true;
   }
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java]: inline `}` on L644: `public final void prepateForProcessing() {}`

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java [hunk 8]
```diff
@@ -503,13 +502,6 @@ public T apply(ValueInSingleWindow<T> input) {
...
   }
 
-  /**
-   * Returns the value of the provided {@link Aggregator}.
-   */
-  public <AggregateT> AggregateT getAggregatorValue(Aggregator<?, AggregateT> agg) {
-    return extractAggregatorValue(agg.getName(), agg.getCombineFn());
-  }
-
   private <AccumT, AggregateT> AggregateT extractAggregatorValue(
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java]: can it be deleted? if not, `@Deprecated` and comment why not.

**Comment by dhalperi** [comment]
Looks great so far!

**Comment by dhalperi** [comment]
Failing due to Javadoc errors in DoFn.

For now, you can just delete the text about createAggregator in DoFn.java and push a fixup to rerun tests.

**Comment by pabloem** [comment]
@dhalperi all tests have run (except for javadoc, flink and apex). Python fails. All else pass.

**Comment by pabloem** [comment]
oops. Misclicked and closed PR. Sorry about that.

**Comment by meunierd** [comment]
I'd appreciate reverting this since it's the only way to instrument streaming jobs in Dataflow.

**Comment by dhalperi** [comment]
Hi Devon,

Unfortunately the removal of Aggregators is a blocker for first stable release of Beam. That metrics are not quite ready yet is true, but this will be fixed before the release. This temporary regression was a choice by the Dataflow owners for the good of the Beam project.
